File: sections/26_tree_dp_basics/problems/d_road_guards/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n = data[0]
    cost = data[1:1 + n]
    graph = [[] for _ in range(n)]
    index = 1 + n
    for _ in range(n - 1):
        u = data[index] - 1
        v = data[index + 1] - 1
        index += 2
        graph[u].append(v)
        graph[v].append(u)

    parent = [-1] * n
    parent[0] = 0
    order = [0]
    for u in order:
        for v in graph[u]:
            if parent[v] == -1:
                parent[v] = u
                order.append(v)

    active = cost[:]
    inactive = [0] * n
    for u in reversed(order):
        for v in graph[u]:
            if parent[v] == u:
                active[u] += min(active[v], inactive[v])
                inactive[u] += active[v]

    print(min(active[0], inactive[0]))


if __name__ == "__main__":
    main()

File: sections/26_tree_dp_basics/problems/d_road_guards/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n)) return 0;
    vector<long long> cost(n);
    for (long long &value : cost) cin >> value;

    vector<vector<int>> graph(n);
    for (int i = 0; i + 1 < n; ++i) {
        int u, v;
        cin >> u >> v;
        --u;
        --v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }

    vector<int> parent(n, -1), order = {0};
    parent[0] = 0;
    for (int index = 0; index < static_cast<int>(order.size()); ++index) {
        int u = order[index];
        for (int v : graph[u]) {
            if (parent[v] == -1) {
                parent[v] = u;
                order.push_back(v);
            }
        }
    }

    vector<long long> active = cost;
    vector<long long> inactive(n, 0);
    for (int index = n - 1; index >= 0; --index) {
        int u = order[index];
        for (int v : graph[u]) {
            if (parent[v] == u) {
                active[u] += min(active[v], inactive[v]);
                inactive[u] += active[v];
            }
        }
    }

    cout << min(active[0], inactive[0]) << '\n';
    return 0;
}
